docs: worker production deployment guide (#19884)
mastra-ai/mastra: docs/src/content/en/docs/deployment/workers.mdx

Reviewer: Now that we have this page we need to make it a little bit more discoverable. In the relevant pages (e.g. workflows overview) we should mention this capability in a short sentence and link to here

---
title: "Workers | Deployment"
description: "Separate background processing from the API layer by running workflow execution, cron schedules, and background tasks in dedicated worker processes."
packages:
  - "@mastra/core"
  - "@mastra/redis-streams"
  - "@mastra/google-cloud-pubsub"
  - "@mastra/pg"
  - "@mastra/libsql"
---

# Workers

Workers handle background processing outside the request-response cycle. Workflow step execution, cron-based scheduling, and long-running tool calls all run in workers, keeping the API responsive.

By default, workers run in the same process as the API. For production workloads, you can split them into separate processes or containers and scale each one independently.

> :::note

Reviewer: This also needs to move to the top below the h1 heading

```suggestion
:::experimental
```

Author: moved this and changed to experimental, thanks

> Workers are a beta feature. The API is stable enough for production use, but some details may change. See [known limitations](#known-limitations) for current gaps.

Reviewer: ```suggestion
This feature is in beta. The API is stable enough for production use, but some details may change. See [known limitations](#known-limitations) for current gaps.
```

Author: Done this, thanks

### Background task worker

Executes agent tool calls marked with `background: { enabled: true }`. When an agent invokes a background tool, the API dispatches the task to this worker instead of blocking the response stream.

The background task worker manages concurrency limits, task lifecycle, and result delivery through the PubSub bus.

## How workers run

### In-process (default)

With no configuration, Mastra creates and starts workers inside the API process. Events flow through an in-memory PubSub, and everything shares a single Node.js runtime.

```typescript title="src/mastra/index.ts"
import { Mastra } from '@mastra/core/mastra'

export const mastra = new Mastra({
  // Workers run in-process by default.
  // No pubsub or worker config needed.
})
```

This setup needs no external infrastructure beyond your storage adapter. It doesn't survive process crashes, and you can't scale individual components.

### Split processes

To run workers separately, configure a distributed [PubSub](/docs/server/pubsub) backend and use the `MASTRA_WORKERS` environment variable to control which workers start in each process.

> The following example uses Redis Streams with PostgreSQL:

Reviewer: For the two examples here use `<Tabs>` component to render them more concisely

export const mastra = new Mastra({
  storage: new PostgresStore({
    connectionString: process.env.DATABASE_URL!,
  }),
  pubsub: new RedisStreamsPubSub({
    url: process.env.REDIS_URL!,
  }),
})
```

The following example uses Google Cloud Pub/Sub with LibSQL:

```typescript title="src/mastra/index.ts"
import { Mastra } from '@mastra/core/mastra'
import { GoogleCloudPubSub } from '@mastra/google-cloud-pubsub'
import { LibSQLStore } from '@mastra/libsql'

export const mastra = new Mastra({
  storage: new LibSQLStore({
    url: process.env.DATABASE_URL!,
  }),
  pubsub: new GoogleCloudPubSub({
    projectId: process.env.GCP_PROJECT_ID!,
  }),
})
```

Any [supported storage backend](#prerequisites-for-split-deployments) works — swap `PostgresStore` or `LibSQLStore` with your preferred backend.

Run the same build artifact in multiple containers, each with a different `MASTRA_WORKERS` value:

| Process | `MASTRA_WORKERS` | Role |
| - | - | - |
| API | `false` | Serves HTTP requests, publishes events, no workers |
| Orchestration | `orchestration` | Pulls workflow events, executes steps |
| Scheduler | `scheduler` | Polls for due schedules, publishes start events |

> | Background tasks | `backgroundTasks` | Executes background tool calls |

Reviewer: This should go into reference docs and then be just linked out in a sentence
